File: A_star.py
# A* on a simple 2D grid (no obstacles)
from queue import PriorityQueue

# Heuristic function: Manhattan Distance
def heuristic(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

# A* algorithm implementation
def a_star(start, goal):
    # Priority Queue to explore the nodes with lowest f_score first
    open_set = PriorityQueue()
    open_set.put((0, start))  # (f_score, node) 
    came_from = {}  # Dictionary to track the parent of each node
    g_score = {start: 0}  # Dictionary to store the cost to reach each node

    while not open_set.empty():
        current = open_set.get()[1]  # Get the node with the lowest f_score
        if current == goal:
            print("Path found!")  # Goal reached
            return  # Return after finding the path
        
        # Neighbors (Up, Right, Down, Left)
        neighbors = [(0,1),(1,0),(0,-1),(-1,0)]
        
        # Explore each neighbor
        for dx, dy in neighbors:
            neighbor = (current[0] + dx, current[1] + dy)
            
            # Ensure the neighbor is within bounds
            if 0 <= neighbor[0] < 5 and 0 <= neighbor[1] < 5:
                temp_g = g_score[current] + 1  # Calculate tentative g_score

                # If the neighbor has not been explored or the new path is shorter
                if neighbor not in g_score or temp_g < g_score[neighbor]:
                    g_score[neighbor] = temp_g  # Update the g_score
                    f_score = temp_g + heuristic(neighbor, goal)  # Calculate f_score
                    open_set.put((f_score, neighbor))  # Add neighbor to priority queue
                    came_from[neighbor] = current  # Track the parent node
    
    print("Path not found!")  # No path found

# Taking user input for start and goal positions
start_x = int(input("Enter the starting X coordinate (0-4): "))
start_y = int(input("Enter the starting Y coordinate (0-4): "))
goal_x = int(input("Enter the goal X coordinate (0-4): "))
goal_y = int(input("Enter the goal Y coordinate (0-4): "))

start = (start_x, start_y)
goal = (goal_x, goal_y)

a_star(start, goal)

'''
............................................................................................
'''
# A* maze solver


import heapq

# Directions: Up, Down, Left, Right
directions = [(-1,0), (1,0), (0,-1), (0,1)]

# Heuristic function: Manhattan Distance
def heuristic(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

# A* maze solver implementation
def astar(maze, start, end):
    # Priority queue to explore the node with lowest f_score
    heap = []
    heapq.heappush(heap, (0 + heuristic(start, end), 0, start, [start]))  # (f_score, g_score, node, path)
    visited = set()  # Set to track visited nodes

    while heap:
        est_total, cost, current, path = heapq.heappop(heap)  # Pop node with lowest f_score

        # If goal is reached, return the path
        if current == end:
            return path

        # Skip already visited nodes
        if current in visited:
            continue
        visited.add(current)

        # Explore neighbors (Up, Down, Left, Right)
        for dx, dy in directions:
            neighbor = (current[0] + dx, current[1] + dy)
            
            # Check if the neighbor is within bounds and not an obstacle
            if (0 <= neighbor[0] < len(maze) and
                0 <= neighbor[1] < len(maze[0]) and
                maze[neighbor[0]][neighbor[1]] == 0):  # 0 indicates an open cell
                # Push the neighbor into the heap with updated g_score and f_score
                heapq.heappush(heap, (
                    cost + 1 + heuristic(neighbor, end),  # f_score
                    cost + 1,  # g_score
                    neighbor,
                    path + [neighbor]  # Add neighbor to the current path
                ))

    return None  # No path found

# Taking user input for the start and end positions
start_x = int(input("Enter the starting X coordinate (0-4): "))
start_y = int(input("Enter the starting Y coordinate (0-4): "))
end_x = int(input("Enter the goal X coordinate (0-4): "))
end_y = int(input("Enter the goal Y coordinate (0-4): "))

start = (start_x, start_y)
end = (end_x, end_y)

# Example maze
maze = [
    [0, 1, 0, 0, 0],
    [0, 1, 0, 1, 0],
    [0, 0, 0, 1, 0],
    [1, 1, 0, 1, 0],
    [0, 0, 0, 0, 0]
]

path = astar(maze, start, end)
if path:
    print("Path found:", path)  # Print the path if found
else:
    print("No path found.")  # No valid path exists

File: A_star.cpp
#include <bits/stdc++.h>
using namespace std;

typedef pair<int, int> pii;

int heuristic(pii a, pii b)
{
    return abs(a.first - b.first) + abs(a.second - b.second);
}

void a_star(pii start, pii goal)
{
    priority_queue<pair<int, pii>, vector<pair<int, pii>>, greater<pair<int, pii>>> open_set;
    open_set.push({0, start});
    map<pii, int> g_score;
    g_score[start] = 0;
    while (!open_set.empty())
    {
        pii current = open_set.top().second;
        open_set.pop();
        if (current == goal)
        {
            cout << "Path found!" << endl;
            return;
        }
        vector<pii> neighbors = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        for (auto [dx, dy] : neighbors)
        {
            pii neighbor = {current.first + dx, current.second + dy};
            if (neighbor.first >= 0 && neighbor.second >= 0 && neighbor.first < 5 && neighbor.second < 5)
            {
                int temp_g = g_score[current] + 1;
                if (!g_score.count(neighbor) || temp_g < g_score[neighbor])
                {
                    g_score[neighbor] = temp_g;
                    int f_score = temp_g + heuristic(neighbor, goal);
                    open_set.push({f_score, neighbor});
                }
            }
        }
    }
    cout << "Path not found!" << endl;
}

int main()
{
    pii start = {0, 0};
    pii goal = {4, 4};
    a_star(start, goal);
    return 0;
}

//......................................................................................//
// A* Maze solver

#include <iostream>
#include <queue>
#include <vector>
#include <tuple>
#include <cmath>
#include <set>
using namespace std;

typedef pair<int, int> pii;

// Directions: up, down, left, right
int dir[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

int heuristic(pii a, pii b)
{
    // Manhattan Distance
    return abs(a.first - b.first) + abs(a.second - b.second);
}

vector<pii> astar(vector<vector<int>> &maze, pii start, pii end)
{
    priority_queue<tuple<int, int, pii, vector<pii>>, vector<tuple<int, int, pii, vector<pii>>>, greater<>> pq;
    set<pii> visited;

    pq.push({heuristic(start, end), 0, start, {start}});

    while (!pq.empty())
    {
        auto [est_total, cost, current, path] = pq.top();
        pq.pop();

        if (current == end)
            return path;

        if (visited.count(current))
            continue;
        visited.insert(current);

        for (auto &d : dir)
        {
            int nx = current.first + d[0];
            int ny = current.second + d[1];

            if (nx >= 0 && ny >= 0 && nx < maze.size() && ny < maze[0].size() && maze[nx][ny] == 0)
            {
                vector<pii> new_path = path;
                new_path.push_back({nx, ny});
                pq.push({cost + 1 + heuristic({nx, ny}, end), cost + 1, {nx, ny}, new_path});
            }
        }
    }
    return {};
}

int main()
{
    vector<vector<int>> maze = {
        {0, 1, 0, 0, 0},
        {0, 1, 0, 1, 0},
        {0, 0, 0, 1, 0},
        {1, 1, 0, 1, 0},
        {0, 0, 0, 0, 0}};

    pii start = {0, 0};
    pii end = {4, 4};

    vector<pii> path = astar(maze, start, end);
    if (!path.empty())
    {
        cout << "Path found: ";
        for (auto [x, y] : path)
            cout << "(" << x << "," << y << ") ";
        cout << endl;
    }
    else
    {
        cout << "No path found." << endl;
    }
}
